zkp/commitments/q_one_way_multiplication.go

```diff
@@ -0,0 +1,206 @@
+package commitmentzkp
+
+import (
+	"github.com/xlab-si/emmy/crypto/common"
+	"math/big"
+)
+
+// ProveCommitmentMultiplication demonstrates how, given commitments A, B, C, prover can
+// prove that C = A * B. Note that commitments need to be based on q-one-way homomorphism
+// (see RSABasedCommitter which is q-one-way homomorphism based).
+func ProveCommitmentMultiplication(homomorphism func(*big.Int) *big.Int, homomorphismInv func(*big.Int) *big.Int,
+	H common.Group, Q, Y,
+	A, B, C, a, b, r, u, o, t *big.Int) bool {
+	prover := NewQOneWayMultiplicationProver(homomorphism, homomorphismInv, H, Q, Y, A, B, C, a, b, r, u, o, t)
+	verifier := NewQOneWayMultiplicationVerifier(homomorphism, H, Q, Y, A, B, C)
+
+	m1, m2, m3 := prover.GetProofRandomData()
+	verifier.SetProofRandomData(m1, m2, m3)
+
+	challenge := verifier.GetChallenge()
+	z1, w1, w2, z2, w3 := prover.GetProofData(challenge)
+	proved := verifier.Verify(z1, w1, w2, z2, w3)
+
+	return proved
+}
+
+type QOneWayMultiplicationProver struct {
+	QOneWayHomomorphism func(*big.Int) *big.Int
+	QOneWayHomomorphismInv func(*big.Int) *big.Int // works only for y^Q, takes y as input
+	H 				 common.Group
+	Q *big.Int
+	Y *big.Int
+	A *big.Int // commitment to a
+	B *big.Int // commitment to b
+	C *big.Int // commitment to c = a * b mod Q
+	a *big.Int
+	b *big.Int
+	r *big.Int // r is random factor in A
+	u *big.Int // u is random factor in B
+	o *big.Int // o is random factor in B
+	t *big.Int // t is such that: C = B^a * f(t)
+	x *big.Int // for protocol 1
```

Comment:
What is `protocol 1` and what is `protocol 2`? Can you please document this in the code?